Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,353 @@
+  ///
+  /// On Windows, this will return the path prefix up to and including the
+  /// host and share for UNC paths or the volume for device paths. This will
+  /// not return any subsequence separator for UNC or device paths.
+  ///
+  /// On Windows, for traditional DOS paths, this will return
+  /// the path prefix up to and including a root directory or
+  /// a supplied drive or volume. Otherwise, if the path is relative to
+  /// both the current directory and current driver, returns `nil`
+  ///
+  /// Examples:
+  /// * Unix:
+  ///   * `/foo/bar => /`
+  ///   * `foo/bar => nil`
+  /// * Windows:
+  ///   * `C:\foo\bar => C:\`
+  ///   * `C:foo\bar => C:`
+  ///   * `\foo\bar => \ `
+  ///   * `foo\bar => nil`
+  ///   * `\\?\UNC\server\share\file => \\?\UNC\server\share`
+  ///
+  /// TODO: Document setter. Better to document for a path what is root and what is relative.
+  ///
+  /// REVIEW NOTE: C# calls this the root. C++ implies they pull in the trailing separator (I
+  /// haven't run tests). Rust differs in that has_root is just looking for the `\` somewhere.
+  ///
+  /// REVIEW NOTE: Going with UNC or device paths not including trailing separator,
+  /// which is consistent with C# IIUC. I need to perform the experiement to verify this.
+  ///
+  /// REVIEW NOTE: I'm going with component as return type, which implies the more common
+  /// use would be to read or inspect the root, rather than forming new paths from the same root.
+  /// For that use case, you can do `FilePath(otherPath.root)`, or use
+  /// the `relativePath`'s setter, or use the `/` operator (if we go that route).
+  public var root: FilePath.Component? {
+    get {
+      let rootEnd = _parseRoot().0
+      guard rootEnd != storage.startIndex else { return nil }
+      return Component(self, ..<rootEnd)
+    }
+    set {
+      fatalError()
+    }
+  }
+
+  /// REVIEW NOTE: I'm dropping rootName and rootDirectory. I think
+  /// C++ was a little misguided here after looking more at C#. I think
+  /// a better alternative would be to actually identify and work with
+  /// the root when on Windows, more akin to Rust's `Prefix` type.
+
+  /// Gets or sets the relative portion of the path (everything after root)
+  ///
+  /// Examples:
+  /// * Unix:
+  ///   * `/foo/bar => foo/bar`
+  ///   * `foo/bar => foo/bar`
+  ///   * `/ => ""`
+  /// * Windows:
+  ///   * `C:\foo\bar => foo\bar`
+  ///   * `foo\bar => foo\bar`
+  ///   * `\\?\UNC\server\share\file => file`
+  ///   * `\ => ""`
+  ///
+  /// REVIEW NOTE: Not optional, can return empty paths. Empty paths
+  /// are relative (but are not roots), hence non-optional.
+  public var relativePath: FilePath {
+    get {
+      let (_, relIdx) = _parseRoot()
+      return FilePath(slice: storage[relIdx...])
+    }
+    set { fatalError() }
+  }
+
+  /// Returns the final component of the path. If the path is a normal file, this
+  /// will be the file name. If the path is a directory, this is the directory name.
```

Comment:
I'm not sure that the mention of a file vs a directory makes sense here, since this is just involved with file path syntax.  Seems to me that at this level of the API, the discussion is just around a sequence of path components, whether or not they even exist in the file system, or whether they are symlinks, special devices, or other things.

Reply:
I was trying to communicate that this was a syntactic consideration and doesn't ask the filesystem whether this is a file or directory, but the wording might be confusing. I'm going with something like this now:

```swift
  /// Returns the final relative component of the path.
  /// Returns `nil` if the path is empty or only contains a root
  ///
  /// Examples:
  /// * Unix:
  ///   * `/usr/local/bin/ => bin`
  ///   * `/tmp/foo.txt    => foo.txt`
  ///   * `/               => nil`
  /// * Windows:
  ///   * `C:\Users\                    => Users`
  ///   * `C:Users\                     => Users`
  ///   * `C:\                          => nil`
  ///   * `\Users\                      => Users`
  ///   * `\\?\UNC\server\share\bar.exe => bar.exe`
  ///   * `\\server\share               => nil`
  ///   * `\\?\UNC\server\share\        => nil`
  ///

```

Reply:
Thanks, that's much clearer.